packages/fhir-router/src/batch.ts

```diff
@@ -22,47 +37,299 @@ import { HttpMethod } from './urlrouter';
+      throw new OperationOutcomeError(badRequest('Transaction requires strict isolation but has too many entries'));
+    }
+
+    return this.repo.withTransaction(() => this.processBatch(bundleInfo, resultEntries), {
+      serializable: bundleInfo.requiresStrongTransaction,
+    });
+  }
+
+  /**
+   * Scans the Bundle in order to ensure entries are processed in the correct sequence,
+   * as well as to identify any operations that might require specific handling.
+   *
+   * @param results - The array of result entries, to track partial results.
+   * @returns The information gathered from scanning the Bundle.
+   */
+  private async preprocessBundle(results: BundleEntry[]): Promise<BundlePreprocessInfo> {
+    const entries = this.bundle.entry;
+    if (!entries?.length) {
+      throw new OperationOutcomeError(badRequest('Missing bundle entries'));
+    }
+
+    const bucketedEntries: Record<RestInteraction, number[]> = {
+      // Processed in order, by interaction type
+      transaction: [],
+      batch: [],
+      delete: [],
+      create: [],
+      update: [],
+      patch: [],
+      operation: [],
+      'search-system': [],
+      'search-type': [],
+      read: [],
+      vread: [],
+      'history-system': [],
+      'history-type': [],
+      'history-instance': [],
+    };
+    const seenIdentities = new Set<string>();
+    let requiresStrongTransaction = false;
+    let updates = 0;
+
+    for (let i = 0; i < entries.length; i++) {
+      const entry = entries[i];
+      const method = entry.request?.method;
+      if (!method) {
+        results[i] = buildBundleResponse(
+          badRequest('Missing Bundle entry request method', `Bundle.entry[${i}].request.method`)
+        );
+        continue;
+      }
+      const outcome = await this.preprocessEntry(entry, i, seenIdentities);
+      if (outcome) {
+        if (!this.isTransaction()) {
+          results[i] = buildBundleResponse(outcome);
+          continue;
+        }
+        throw new OperationOutcomeError(outcome);
+      }
+
+      const route = this.getRouteForEntry(entry);
+      const interaction = route?.data?.interaction as RestInteraction;
+      if (interaction === 'create' && entry.request?.ifNoneExist) {
+        // Conditional create requires strong (serializable) transaction to
+        // guarantee uniqueness of created resource
+        requiresStrongTransaction = true;
+      } else if (interaction === 'update') {
+        if (entry.request?.url.includes('?')) {
+          // Conditional update requires strong (serializable) transaction to
+          // guarantee uniqueness of possibly-created resource
+          requiresStrongTransaction = true;
+        }
+        updates++;
+      } else if (interaction === 'delete' && entry.request?.url.includes('?')) {
+        // Conditional delete requires strong (serializable) transaction
+        requiresStrongTransaction = true;
+      }
+      bucketedEntries[interaction]?.push(i);
```

Comment:
Can we be sure that `interaction` is valid here?

Reply:
I'll add a guard above to ensure that it is